[BUG] Regression 3.0.12 -> 3.1.0+  generated source contains duplicate definitions

### Describe the bug

Here's the console output from manually bisecting the PyPI releases:
```console
(cpython-v3.13-apt-deb) 20:47 root@argus:/work/external/chemreac# py3 -m pip install cython==3.0.12
Collecting cython==3.0.12
  Downloading Cython-3.0.12-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.5 kB)
Downloading Cython-3.0.12-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 71.0 MB/s eta 0:00:00
Installing collected packages: cython
  Attempting uninstall: cython
    Found existing installation: Cython 3.1.0
    Uninstalling Cython-3.1.0:
      Successfully uninstalled Cython-3.1.0
Successfully installed cython-3.0.12
(cpython-v3.13-apt-deb) 20:47 root@argus:/work/external/chemreac# PYTHONPATH=/work/external/pycvodes:$PYTHONPATH env -C /work/external/chemreac     CC="ccache $CC"     PYCVODES_NO_LAPACK=${PYCVODE_NO_LAPACK:-"0"}     PYCVODES_NO_KLU=${PYCVODE_NO_KLU:-"0"}     CFLAGS="$PYODESYS_CVODE_FLAGS"     CXXFLAGS="$PYODESYS_CVODE_FLAGS"     LDFLAGS="$PYODESYS_CVODE_LDFLAGS"     py3 setup.py build_ext -i
/work/external/chemreac/setup.py:61: UserWarning: Using git to derive version: dev-branches may compete.
  warnings.warn("Using git to derive version: dev-branches may compete.")
cythonize_inc_path=['chemreac/include', '/work/external/pycvodes/pycvodes/include', 'external/anyode/cython_def']
Compiling ./chemreac/_chemreac.pyx because it changed.
[1/1] Cythonizing ./chemreac/_chemreac.pyx
/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13/site-packages/finitediff/__init__.py:23: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import resource_filename, Requirement
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/setuptools/dist.py:332: InformationOnly: Normalizing '0.8.3.post69+g2cf8e08-dirty' to '0.8.3.post69+g2cf8e08.dirty'  self.metadata.version = self._normalize_version(self.metadata.version)
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
!!

        ********************************************************************************
        Please consider removing the following classifiers in favor of a SPDX license expression:

        License :: OSI Approved :: BSD License

        See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
        ********************************************************************************

!!
  self._finalize_license_expression()
running build_ext
building 'chemreac._chemreac' extension
g++-14 -isystem /opt-3/sundials-6.7.0-release/include -fPIC -I/work/external/pycvodes/pycvodes/include -Ichemreac/include -I/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-
packages/numpy/_core/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13/site-packages/finitediff/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13
/site-packages/block_diag_ilu/include -I/work/external/pycvodes/pycvodes/include -Ichemreac/include -Iexternal/anyode/include -I/opt-3/cpython-v3.13-apt-deb/include -I/usr/i
nclude/python3.13 -c ./chemreac/_chemreac.cpp -o build/temp.linux-x86_64-cpython-313/chemreac/_chemreac.o -std=c++20
In file included from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/ndarraytypes.h:1913,
                 from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/ndarrayobject.h:12,
                 from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/arrayobject.h:5,
                 from ./chemreac/_chemreac.cpp:1262:
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it 
with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
g++-14 -isystem /opt-3/sundials-6.7.0-release/include -fPIC -I/work/external/pycvodes/pycvodes/include -Ichemreac/include -I/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-
packages/numpy/_core/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13/site-packages/finitediff/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13
/site-packages/block_diag_ilu/include -I/work/external/pycvodes/pycvodes/include -Ichemreac/include -Iexternal/anyode/include -I/opt-3/cpython-v3.13-apt-deb/include -I/usr/i
nclude/python3.13 -c src/chemreac.cpp -o build/temp.linux-x86_64-cpython-313/src/chemreac.o -std=c++20
g++-14 -isystem /opt-3/sundials-6.7.0-release/include -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,--disable-new-dtags -Wl,-rpath,/opt-3/sundials
-6.7.0-release/lib -L/opt-3/sundials-6.7.0-release/lib -isystem /opt-3/sundials-6.7.0-release/include build/temp.linux-x86_64-cpython-313/chemreac/_chemreac.o build/temp.lin
ux-x86_64-cpython-313/src/chemreac.o -L/usr/lib/x86_64-linux-gnu -lsundials_nvecserial -lsundials_cvodes -lsundials_sunlinsolspgmr -lsundials_sunlinsolspbcgs -lsundials_sunl
insolsptfqmr -lsundials_sunmatrixdense -lsundials_sunmatrixband -lsundials_sunlinsollapackdense -lsundials_sunlinsollapackband -lsundials_sunlinsolklu -llapack -lblas -lm -o
 build/lib.linux-x86_64-cpython-313/chemreac/_chemreac.cpython-313-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-cpython-313/chemreac/_chemreac.cpython-313-x86_64-linux-gnu.so -> chemreac
(cpython-v3.13-apt-deb) 20:47 root@argus:/work/external/chemreac# py3 -m pip install cython==3.1.0
Collecting cython==3.1.0
  Using cached cython-3.1.0-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (30 kB)
Using cached cython-3.1.0-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)
Installing collected packages: cython
  Attempting uninstall: cython
    Found existing installation: Cython 3.0.12
    Uninstalling Cython-3.0.12:
      Successfully uninstalled Cython-3.0.12
Successfully installed cython-3.1.0
(cpython-v3.13-apt-deb) 20:48 root@argus:/work/external/chemreac# PYTHONPATH=/work/external/pycvodes:$PYTHONPATH env -C /work/external/chemreac     CC="ccache $CC"     PYCVO
DES_NO_LAPACK=${PYCVODE_NO_LAPACK:-"0"}     PYCVODES_NO_KLU=${PYCVODE_NO_KLU:-"0"}     CFLAGS="$PYODESYS_CVODE_FLAGS"     CXXFLAGS="$PYODESYS_CVODE_FLAGS"     LDFLAGS="$PYOD
ESYS_CVODE_LDFLAGS"     py3 setup.py build_ext -i
/work/external/chemreac/setup.py:61: UserWarning: Using git to derive version: dev-branches may compete.
  warnings.warn("Using git to derive version: dev-branches may compete.")
cythonize_inc_path=['chemreac/include', '/work/external/pycvodes/pycvodes/include', 'external/anyode/cython_def']
Compiling ./chemreac/_chemreac.pyx because it changed.
[1/1] Cythonizing ./chemreac/_chemreac.pyx
/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13/site-packages/finitediff/__init__.py:23: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pyp
a.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import resource_filename, Requirement
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/setuptools/dist.py:332: InformationOnly: Normalizing '0.8.3.post69+g2cf8e08-dirty' to '0.8.3.post69+g2cf8e08.dirty'
  self.metadata.version = self._normalize_version(self.metadata.version)
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
!!

        ********************************************************************************
        Please consider removing the following classifiers in favor of a SPDX license expression:

        License :: OSI Approved :: BSD License

        See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
        ********************************************************************************

!!
  self._finalize_license_expression()
running build_ext
building 'chemreac._chemreac' extension
g++-14 -isystem /opt-3/sundials-6.7.0-release/include -fPIC -Ichemreac/include -I/work/external/pycvodes/pycvodes/include -I/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-
packages/numpy/_core/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13/site-packages/finitediff/include -I/work/.ci/build-rel-3.13.3/py-usr-site/lib/python3.13
/site-packages/block_diag_ilu/include -I/work/external/pycvodes/pycvodes/include -Ichemreac/include -Iexternal/anyode/include -I/opt-3/cpython-v3.13-apt-deb/include -I/usr/i
nclude/python3.13 -c ./chemreac/_chemreac.cpp -o build/temp.linux-x86_64-cpython-313/chemreac/_chemreac.o -std=c++20
In file included from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/ndarraytypes.h:1913,
                 from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/ndarrayobject.h:12,
                 from /opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/arrayobject.h:5,
                 from ./chemreac/_chemreac.cpp:1150:
/opt-3/cpython-v3.13-apt-deb/lib/python3.13/site-packages/numpy/_core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it 
with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
./chemreac/_chemreac.cpp:5190:18: error: redefinition of ‘PyObject* __pyx_convert_vector_to_py_double(const std::vector<double>&)’
 5190 | static PyObject *__pyx_convert_vector_to_py_double(std::vector<double>  const &__pyx_v_v) {
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./chemreac/_chemreac.cpp:4764:18: note: ‘PyObject* __pyx_convert_vector_to_py_double(const std::vector<double>&)’ previously defined here
 4764 | static PyObject *__pyx_convert_vector_to_py_double(std::vector<double>  const &__pyx_v_v) {
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: command '/usr/bin/g++-14' failed with exit code 1
```

### Code to reproduce the behaviour:

I've uploaded the problematic generated .cpp file and the original .pyx file to a gist:
https://gist.github.com/bjodah/cac706265f270a820cdc0ff91f4364b4

### Expected behaviour

_No response_

### OS

_No response_

### Python version

3.13

### Cython version

3.0.12 (ok), 3.1.0+ (failing)

### Additional context

_No response_

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -4248,8 +4248,12 @@ class CppClassType(CType):
                 'type': self.cname,
             })
             # Override directives that should not be inherited from user code.
+            # Use the module-level (global scope) directives to ensure that the same
+            # conversion utility is not emitted twice when it is requested from two
+            # different local scopes that carry different directive values for settings
+            # that are irrelevant to the generated C helper (e.g. 'binding').
             from .UtilityCode import CythonUtilityCode
-            directives = CythonUtilityCode.filter_inherited_directives(env.directives)
+            directives = CythonUtilityCode.filter_inherited_directives(env.global_scope().directives)
             env.use_utility_code(CythonUtilityCode.load(
                 cls.replace('unordered_', '') + ".from_py", "CppConvert.pyx",
                 context=context, compiler_directives=directives))
@@ -4294,8 +4298,11 @@ class CppClassType(CType):
                 'type': self.cname,
             })
             from .UtilityCode import CythonUtilityCode
-            # Override directives that should not be inherited from user code.
-            directives = CythonUtilityCode.filter_inherited_directives(env.directives)
+            # Use the module-level (global scope) directives to ensure that the same
+            # conversion utility is not emitted twice when it is requested from two
+            # different local scopes that carry different directive values for settings
+            # that are irrelevant to the generated C helper (e.g. 'binding').
+            directives = CythonUtilityCode.filter_inherited_directives(env.global_scope().directives)
             env.use_utility_code(CythonUtilityCode.load(
                 cls.replace('unordered_', '') + ".to_py", "CppConvert.pyx",
                 context=context, compiler_directives=directives))

--- a/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
+++ b/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
@@ -1,5 +1,5 @@
 # mode: compile
-# tag: cpp no-cpp-locals
+# tag: cpp, no-cpp-locals
 
 # Regression test for gh-6981: duplicate __pyx_convert_vector_to_py_double in generated C++.
 # Trigger requires both:

--- a/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
+++ b/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
@@ -1,5 +1,5 @@
 # mode: compile
-# tag: cpp
+# tag: cpp no-cpp-locals
 
 # Regression test for gh-6981: duplicate __pyx_convert_vector_to_py_double in generated C++.
 # Trigger requires both:

--- a/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
+++ b/tests/compile/cpp_vector_to_py_dedup_T6981.pyx
@@ -1,0 +1,26 @@
+# mode: compile
+# tag: cpp
+
+# Regression test for gh-6981: duplicate __pyx_convert_vector_to_py_double in generated C++.
+# Trigger requires both:
+#   (a) an extension-type property returning a vector[T] member of a template cppclass, and
+#   (b) a module-level def with a vector[T] default argument.
+
+from libcpp.vector cimport vector
+
+cdef extern from *:
+    """
+    #include <vector>
+    template<class T> struct RD { std::vector<T> v; };
+    """
+    cdef cppclass RD[T]:
+        vector[T] v
+
+cdef class C:
+    cdef RD[double] *p
+    property v:
+        def __get__(self):
+            return self.p.v
+
+def f(vector[double] x=[]):
+    pass

